<!DOCTYPE html>
<html lang="en" ng-app="tevApp">
<head>
    <meta charset="UTF-8">
    <title>TEV</title>
</head>
<style>
    body {
        padding-top: 75px; /* Required padding for .navbar-fixed-top. Remove if using .navbar-static-top. Change if height of navigation changes. */
    }

    /* bootstrap hack: fix content width inside hidden tabs */
    .tab-content > .tab-pane:not(.active),
    .pill-content > .pill-pane:not(.active) {
        display: block;
        height: 0;
        overflow-y: hidden;
    }

    .title-text {
        font-family: Arial, Helvetica, sans-serif;
        font-size: large;
        text-decoration: underline;
    }

    tr [colspan="2"] {
        text-align: center;
    }

    /* bootstrap hack end */

    .axis path, .axis line {
        fill: none;
        stroke: #000;
        shape-rendering: crispEdges;
    }

    .purity-line-ticks {
        font-size: 10px;
    }

    .node circle {
        stroke-width: 1.5px;
    }

    .node_text {
        font: 10px sans-serif;
    }

    .link {
        fill: none;
        stroke: #ccc;
        stroke-width: 1.5px;
    }

    .info_rects, .median_line:hover {
        cursor: pointer;
    }

    div.tooltip {
        position: absolute;
        text-align: center;
        padding: 2px;
        font: 12px sans-serif;
        background: #222;
        border: 2px;
        border-color: #222222;
        border-radius: 8px;
    }

    /* Hide arrows in number inputs */
    input[type=number]::-webkit-inner-spin-button,
    input[type=number]::-webkit-outer-spin-button {
        -webkit-appearance: none;
        -moz-appearance: none;
        appearance: none;
        margin: 0;
    }

    /* When mini fishplot is being dragged (mousedown -> mousemove), prevent any other mouse operations */
    .noselect {
        -webkit-touch-callout: none;
        -webkit-user-select: none;
        -khtml-user-select: none;
        -moz-user-select: none;
        -ms-user-select: none;
        user-select: none;
    }

    .modal-header-inverse {
        background: #222222;
        color: #ffffff;
    }

    .panel-default > .panel-heading-custom {
        background: #222222;
        color: #ffffff;
    }

</style>

<script src="https://ajax.googleapis.com/ajax/libs/angularjs/1.5.5/angular.js"></script>
<script src="https://d3js.org/d3.v3.min.js" charset="utf-8"></script>


<script src="https://ajax.googleapis.com/ajax/libs/jquery/2.2.4/jquery.min.js"></script>

{% load static from staticfiles %}
<script src="{% static "js/tev_helper_functions.js" %}" type="text/javascript"></script>
<script src="{% static "js/figure1plot.js" %}" type="text/javascript"></script>
<script src="{% static "js/figure2plot.js" %}" type="text/javascript"></script>
<script src="{% static "js/figure5plot.js" %}" type="text/javascript"></script>
<script src="{% static "js/fishplot_editor.js" %}" type="text/javascript"></script>
<script src="{% static "js/purityplot.js" %}" type="text/javascript"></script>

<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css">

<!-- Optional theme -->
<link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.6/css/bootstrap-theme.min.css"
      integrity="sha384-fLW2N01lMqjakBkx3l/M9EahuwpSfeNvV63J5ezn3uZzapT0u7EYsXMjQV+0En5r"
      crossorigin="anonymous">

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js"></script>
<script src="https://use.fontawesome.com/6138a840c3.js"></script>

<script>

    (function() {
        //Application name: tevApp
        //Dependencies: None
        var tevApp = angular.module('tevApp', []);

        //To distinguish between Django variables used in HTML
        tevApp.config(function($interpolateProvider) {
            $interpolateProvider.startSymbol('[{');
            $interpolateProvider.endSymbol('}]');
        });

        tevApp.config(function ($httpProvider) {
            $httpProvider.defaults.xsrfCookieName = 'csrftoken';
            $httpProvider.defaults.xsrfHeaderName = 'X-CSRFToken';
        });

        tevApp.factory('alleleify', ['$http', function($http){
            return {
                alleleify: function(data, gene_data){
                    var allele_data = [];

                    //Loop through to get all uuid
                    //So we only have to access them, instead of loop through for every amino acid change
                    var gene_uuids = [];
                    var chromosomes = {};
                    var positions = {};
                    var types = {};
                    var cDNAs = {};
                    var ref_seqs = {};
                    var references = {};
                    var alternatives = {};

                    for(var i = 0; i < gene_data.length; i++){
                        gene_uuids.push(gene_data[i].uuid);
                    }

                    for(var i = 0; i < data.length; i++){
                        var timepoint = data[i].timepoint;
                        var alleles = data[i].VariantAlleles;
                        for(var j = 0; j < alleles.length; j++){
                            var allele = gene_data[gene_uuids.indexOf(alleles[j].gene)].name
                                    + '-'
                                    + alleles[j].AA_original
                                    + alleles[j].AA_position.toString()
                                    + alleles[j].AA_variant;
                            var alt_count = alleles[j].alternative_freq;
                            if(chromosomes[allele] == undefined){
                                chromosomes[allele] = gene_data[gene_uuids.indexOf(alleles[j].gene)].chromosome;
                                positions[allele] = gene_data[gene_uuids.indexOf(alleles[j].gene)].position;
                                types[allele] = alleles[j].type;
                                cDNAs[allele] = alleles[j].cDNA_change;
                                ref_seqs[allele] = alleles[j].ref_seq;
                                references[allele] = alleles[j].reference;
                                alternatives[allele] = alleles[j].alternative;
                            }
                            allele_data.push({
                                allele: allele,
                                Sample_Barcode: timepoint,
                                alt_count: alt_count
                            });
                        }
                    }
                    return [allele_data, chromosomes, positions, types, cDNAs, ref_seqs, references, alternatives];
                },

                patch: function(data){
                    var timepoints = uniqueAxisLabels(data, 'Sample_Barcode');
                    data.sort(function(a,b) {return (a.allele > b.allele) ? 1 : ((b.allele > a.allele) ? -1 : 0);} );

                    var current_allele = data[0].allele;
                    var timepoints_hit = [];
                    var unhit_array = [];

                    for(var i = 0; i < data.length; i++){
                        if(data[i].allele == current_allele){
                            timepoints_hit.push(data[i].Sample_Barcode);
                        }
                        else{
                            var unhit_timepoints = timepoints.diff(timepoints_hit);
                            for(var j = 0; j < unhit_timepoints.length; j++){
                                unhit_array.push({
                                    allele: current_allele,
                                    Sample_Barcode: unhit_timepoints[j],
                                    alt_count: 0
                                });
                            }
                            timepoints_hit = [];
                            current_allele = data[i].allele;
                            timepoints_hit.push(data[i].Sample_Barcode);
                        }
                    }
                    //Get the unhit timepoints from last remaining allele
                    unhit_timepoints = timepoints.diff(timepoints_hit);
                    for(var j = 0; j < unhit_timepoints.length; j++){
                                unhit_array.push({
                                    allele: current_allele,
                                    Sample_Barcode: unhit_timepoints[j],
                                    alt_count: 0
                                });
                            }

                    data = data.concat(unhit_array);
                    return data;
                },

                allele_color_reference: function(alleleified_data){
                    var colors = ['#1f77b4', '#ff7f0e', '#2ca02c', '#d62728', '#9467bd', '#8c564b', '#e377c2', '#bcbd22', '#17becf',
                    '#393b79', '#637939', '#b5cf6b', '#e7ba52', '#d6616b', '#843c39', '#7b4173'];

                    var nested_data = d3.nest().key(function (d) {
                                return d.allele;
                            })
                            .entries(alleleified_data);

                    var color_ref = {};
                    for(var i = 0; i < nested_data.length; i++){
                        var allele = nested_data[i].key;
                        color_ref[allele] = colors[i];
                    }
                    return color_ref;
                },

                patients_with_gene: function (alleleified_data, source_data, gene_data) {
                    //nest alleles to get gene names, will use gene_data to convert uuid to these gene names when done
                    var alleles = d3.nest().key(function (d) {
                                return d.allele;
                            })
                            .entries(alleleified_data);
                    var genes = [];
                    var gene_to_uuid = {};
                    var gene_uuids = [];
                    var genes_dict = {};
                    for (var i = 0; i < alleles.length; i++) {
                        genes.push(alleles[i].key.split('-')[0]);
                    }
                    //grab gene uuids to look up in sources (foreign key)
                    for (var i = 0; i < gene_data.length; i++) {
                        if (genes.indexOf(gene_data[i].name) != -1) {
                            genes_dict[gene_data[i].uuid] = [];
                            gene_uuids.push(gene_data[i].uuid);
                            gene_to_uuid[gene_data[i].name] = gene_data[i].uuid;
                        }
                    }

                    for (var i = 0; i < source_data.length; i++) {
                        var source = source_data[i].subject_id;
                        var samples = source_data[i].Samples;
                        var genes_used = [];
                        for (var j = 0; j < samples.length; j++) {
                            var variants = samples[j].VariantAlleles;
                            for (var k = 0; k < variants.length; k++) {
                                if (gene_uuids.indexOf(variants[k].gene) != -1 && genes_used.indexOf(variants[k].gene) == -1) {
                                    genes_used.push(variants[k].gene);
                                    genes_dict[variants[k].gene].push(source);
                                }
                                if (genes_used.length == gene_uuids.length) {
                                    break;
                                }
                            }
                            if (genes_used.length == gene_uuids.length) {
                                break;
                            }
                        }
                    }

                    for(var i = 0; i < genes.length; i++){
                        genes_dict[genes[i]] = genes_dict[gene_to_uuid[genes[i]]];
                    }
                    return genes_dict;
                }
            };
        }]);

        tevApp.factory('saved_fishplots', function(){
           return {
              get_names: function (name_data){
                  var names_array = [];
                  name_data =  name_data.saved_as;

                  for(var i = 0; i < name_data.length; i++){
                      names_array.push(name_data[i].name);
                  }
                  return names_array;
              }
           } ;
        });

        tevApp.factory('purity', function(){
           return {
               init_purities: function(data){
                   var timepoints = uniqueAxisLabels(data, 'Sample_Barcode');
                   var purity_dict = {};
                   for(var i = 0; i < timepoints.length; i++){
                       purity_dict[timepoints[i]] = 90;
                   }
                   return purity_dict;
               }
           }
        });


        //Controller name: TevDataController
        //Services: scope, http
        tevApp.controller('tevDataController', ['$scope', '$http', '$filter', 'alleleify', '$q', 'saved_fishplots', 'purity', '$window',
            function($scope, $http, $filter, allelify, $q, saved_fishplots, purity, $window){
            //Initialize the scopes data to empty (will render blank plots until we've fetched data)

                angular.element($window).on('resize', $scope.$apply.bind($scope));

                $scope.tevData = [];
                $scope.alleleified_data = [];
                //parameter for editor table only, when user loads previously saved fishplot
                $scope.loaded_fishplot_data = [];
                $scope.chromosome_dict = {};
                $scope.position_dict = {};
                $scope.type_dict = {};
                $scope.cDNA_dict = {};
                $scope.ref_seq_dict = {};
                $scope.reference_dict = {};
                $scope.alternative = {};
                $scope.linked_alleleified_data = [];
                $scope.dendro_data = [];
                $scope.dendro_y_data = [];
                $scope.fishbone_data = [];
                //Created this variable because if alleleified_data is changed by dendrogram grouping,
                // it would keep calling itself
                $scope.clustered_alleleified_data = [];
                $scope.alleles = [];
                $scope.subject_id = '';
                $scope.saved_fishplot_names = [];
                $scope.updated_editor_info = [];
                $scope.color_ref = {};
                //dictionary with how many patients in DB share mutation in gene
                //could do on allele basis?
                $scope.patient_gene_dict = {};

                //data for purity line plot above fishplot editor
                $scope.purity_data = {};

                //Check if an allele has already been plotted. For editor table... dont allow change if plotted
                $scope.plotted = [];
                $scope.prompt = 0;

                var source_uuid = '{{ source }}';
                var dendro_data = {{ d3_dendro_data|safe }};
                var dendro_y_data = {{ dendro_y_data }};

                //Get the data from the API
                var source_request = $http.get('../api/v1/sources/');
                var gene_request = $http.get('../api/v1/genes/');
                var sample_request = $http.get('../api/v1/samples/');
                var saved_fishplot_names_request = $http.get('../api/v1/saved_fishplots_just_names/');

                $q.all([source_request, gene_request, saved_fishplot_names_request]).then(function(results_array){
                    var tevData = results_array[0].data;
                    var source_data = results_array[0].data;
                    var gene_data = results_array[1].data;
                    var saved_fishplot_names = results_array[2].data;
                    tevData = $filter('filter')(tevData, {'uuid': source_uuid})[0];
                    $scope.subject_id = tevData.subject_id;
                    saved_fishplot_names = $filter('filter')(saved_fishplot_names, {'subject_id': tevData.subject_id})[0];
                    if(saved_fishplot_names != undefined){
                        saved_fishplot_names = saved_fishplots.get_names(saved_fishplot_names);
                        $scope.saved_fishplot_names = saved_fishplot_names;
                    }

                    d3.select('#subject_brand').html(function(){
                        return $scope.subject_id;
                    });
                    $scope.tevData = tevData.Samples;

                    var alleleified_data = allelify.alleleify($scope.tevData, gene_data);
                    $scope.chromosome_dict = alleleified_data[1];
                    $scope.position_dict = alleleified_data[2];
                    $scope.type_dict = alleleified_data[3];
                    $scope.cDNA_dict = alleleified_data[4];
                    $scope.ref_seq_dict = alleleified_data[5];
                    $scope.reference_dict = alleleified_data[6];
                    $scope.alternative_dict = alleleified_data[7];
                    alleleified_data = alleleified_data[0];
                    alleleified_data = allelify.patch(alleleified_data);
                    $scope.alleleified_data = alleleified_data;
                    $scope.color_ref = allelify.allele_color_reference(alleleified_data);
                    $scope.patient_gene_dict = allelify.patients_with_gene(alleleified_data, source_data, gene_data);

                    $scope.purity_data = purity.init_purities(alleleified_data);

                    //Send through this data, when the data from the API has been recieved
                    $scope.dendro_data = dendro_data;
                    $scope.dendro_y_data = dendro_y_data;
                });
        }]);

        //Directive that contains d3 code to render allele frequency change over timepoints
        //Plot type: Line plot
        tevApp.directive('figure1Plot',['$window', function($window){
            return {
               scope: {
                   data: '=',
                   colors: '='
               },
               restrict: 'A',
               link: link
           };

            function link(scope, element){
                var data = scope.data;
                var el = element[0];
                var width = el.clientWidth;
                var height = el.clientHeight;

                var color_ref = scope.colors;

                scope.$watch('colors', function(new_colors){
                   color_ref = new_colors;
                });

                scope.$watch('data', function(updated_data){
                    overtimeLinePlot(updated_data, el, height, width, color_ref);

                    $window.addEventListener('resize', overtime_resize);
                    function overtime_resize() {
                        var new_height = d3.select(el).node().clientHeight;
                        var new_width = d3.select(el).node().clientWidth;
                        overtimeLinePlot(updated_data, el, new_height, new_width, color_ref);
                    }
                }, true);
            }
        }]);

        //Directive that contains d3 code to render distribution of allele frequency at timepoints
        //Plot type: Box plots
        tevApp.directive('figure2Plot',['$window', function($window){
            return{
                restrict: 'A',
                link: link,
                scope: {
                    data: '=',
                    colors: '='
                }
            };

            function link(scope, element){
                var data = scope.data[0];
                var alleles = scope.data[1];
                var el = element[0];
                var width = el.clientWidth;
                var height = el.clientHeight;

                var color_ref = scope.colors;
                var current_extension = d3.select(el).attr('ext', 0);


                scope.$watch('colors', function(new_colors){
                    color_ref = new_colors;
                });

                scope.$watch('data', function(updatedData){
                    makeBoxPlots(updatedData, el, height, width, color_ref);

                    $window.addEventListener('resize', boxplot_resize);
                    function boxplot_resize(){
                        var new_width = d3.select(el).node().clientWidth;
                        makeBoxPlots(updatedData, el, height, new_width, color_ref);
                    }
                }, true);
            }
        }]);

        //Directive that contains d3 code to render dendrogram clustered by alternative allele freq
        //Plot type: Elbow dendrogram
        tevApp.directive('figure3Plot', function(){

            return {
               restrict: 'A',
               scope: {
                   data: '=',
                   colors: '='
               },
               link: link
            };

            function link(scope, element){
                var data = scope.data[0];
                var dendro_y_positions = scope.data[1];
                var alleleified_data = scope.data[2];
                var el = element[0];
                var width = el.clientWidth;
                var height = el.clientHeight;

                var color_ref = scope.data[3];

                var cluster = d3.layout.cluster()
                        .size([420, 420]);

                scope.$watch('data', function(updatedData){
                    d3.select(el).select('svg').remove();
                    var alleles = [];

                    data = updatedData[0];
                    dendro_y_positions = updatedData[1];
                    alleleified_data = updatedData[2];

                    color_ref = updatedData[3];

                    if(alleleified_data.length == 0){
                        return;
                    }

                    var canvas = d3.select(el).append('svg')
                            .attr('height', height)
                            .attr('width', width);

                    var title = canvas.append('text')
                            .attr('class', 'title-text')
                            .attr('x', width/2)
                            .attr('y', 20)
                            .style('text-anchor', 'middle')
                            .text(function(){
                                return 'Clustering of Alleles by VAF';
                            });

                    var sliderSvg = canvas.append('g');

                    canvas = canvas.append('g')
                            .attr('transform', 'translate(0, 75)');

                    var nodes = cluster.nodes(data);
                    var links = cluster.links(nodes);

                    //Update dendrogram position based on python clustering position
                    nodes.forEach(function(d, i){
                        d.y = dendro_y_positions[i];
                        d.id = i;
                    });

                    var drag = d3.behavior.drag()
                                .on('drag', function () {
                                    var slider = d3.select(this);
                                    var max_path_y = yScale.invert(30);
                                    if (d3.event.x < 30) {
                                        slider.attr('cx', 30);
                                        d3.select('#cutoff_line').attr('x1', 30).attr('x2', 30);
                                        max_path_y = yScale.invert(30);
                                    }
                                    else if (d3.event.x > 450) {
                                        slider.attr('cx', 450);
                                        d3.select('#cutoff_line').attr('x1', 450).attr('x2', 450);
                                        max_path_y = yScale.invert(450);
                                    }
                                    else {
                                        slider.attr('cx', d3.event.x);
                                        var cutoff = d3.select('#cutoff_line').attr('x1', d3.event.x).attr('x2', d3.event.x);
                                        max_path_y = yScale.invert(d3.event.x);
                                    }
                                    var linked_alleleified_data = cut_data(max_path_y, alleleified_data, nodes, scope.$parent.tevData);
                                    var clusters = linked_alleleified_data[2];
                                    var parents = linked_alleleified_data[3];

                                    for(var i = 0; i < clusters.length; i++){
                                        var alleles_in_cluster = clusters[i];
                                        for(var j = 0; j < alleles_in_cluster.length; j++){
                                            var id = '#allele' + alleles.indexOf(alleles_in_cluster[j]);
                                            var node = d3.select(id);
                                            for(var k = 0; k < parents.length; k++) {
                                                if(alleles_in_cluster.indexOf(parents[k]) != -1) {
                                                    node.attr('fill', color_ref[parents[k]]);
                                                }
                                            }
                                        }
                                    }
                                    scope.$apply(function(){
                                        scope.$parent.linked_alleleified_data = linked_alleleified_data[0];
                                        scope.$parent.clustered_alleleified_data = linked_alleleified_data[1];
                                    });
                                });

                    sliderSvg.append('line')
                            .attr('x1', 30)
                            .attr('x2', 450)
                            .attr('y1', 50)
                            .attr('y2', 50)
                            .attr('stroke', 'black')
                            .attr('stroke-width', 1.5)
                            .attr('fill', 'none');

                    sliderSvg.append('circle')
                            .attr('r', 7)
                            .attr('cy', 50)
                            .attr('cx', 450)
                            .attr('id', 'slider_circle')
                            .attr('stroke', 'black')
                            .attr('stroke-width', 0.5)
                            .attr('fill', '#EEEEEE')
                            .style('cursor', 'pointer')
                            .call(drag);

                    var y_max = Math.max.apply(null, dendro_y_positions);
                    var yScale = d3.scale.linear().domain([y_max, 0]).range([30, 450]);
                    var yAxis = d3.svg.axis().scale(yScale).orient('top');

                    var link = canvas.selectAll('.link')
                            .data(links)
                            .enter()
                            .append('path')
                            .attr('class', 'link')
                            .attr('d', function elbow(d, i) {
                                return "M" + yScale(d.source.y) + "," + d.source.x
                                        + "V" + d.target.x + "H" + yScale(d.target.y);
                            })
                            .attr('fill', 'none')
                            .attr('stroke', "#ccc")
                            .attr('stroke-width', 1.5)
                            .attr('transform', 'translate(0,40)');

                    var node = canvas.selectAll('.node')
                            .data(nodes)
                            .enter()
                            .append('g')
                            .attr('transform', function(d){return "translate("+ yScale(d.y) + "," + (d.x + 40) + ")";});

                    node.append('circle')
                            .attr('r', function(d){
                                if(d.name == " "){
                                    return 0;
                                }
                                return 4.5;
                            })
                            .attr('id', function(d){
                                if(d.name != " ") {
                                    alleles.push(d.name);
                                    return 'allele'+ (d.cluster-1);
                                }
                                else{
                                    return "empty";
                                }
                            })
                            .attr('stroke-width', 1.5)
                            .attr('fill', function(d){
                                if(d3.select(this).attr('id') == 'empty'){
                                    return;
                                }
                                return color_ref[d.parent_allele];
                            });

                    canvas.append('line')
                            .attr('x1', function(){
                                var x = d3.select('#slider_circle').attr('cx');
                                return x;
                            })
                            .attr('x2', function(){
                                var x = d3.select('#slider_circle').attr('cx');
                                return x;
                            })
                            .attr('y1', 30)
                            .attr('y2', 450)
                            .attr('id', 'cutoff_line')
                            .attr('stroke', 'black')
                            .attr('stroke-width', 1)
                            .attr('fill', 'none');

                    node.append("text")
                            .attr("dx", function(d) { return d.children ? -8 : 8; })
                            .attr("dy", 3)
                            .attr('class', 'node_text')
                            .attr("text-anchor", function(d) { return d.children ? "end" : "start"; })
                            .text(function(d) { return d.name; });

                    canvas.append("svg:g")
                            .attr("class", "axis")
                            .attr('transform', 'translate(0, 25)')
                            .call(yAxis);

                    //Initialize the fishplot data
                    var linked_alleleified_data = cut_data(0, alleleified_data, nodes, scope.$parent.tevData);
                    scope.$parent.linked_alleleified_data = linked_alleleified_data[0];
                    scope.$parent.clustered_alleleified_data = linked_alleleified_data[1];
                    //Input the alleles, ordered by clustering
                    scope.$parent.alleles = alleles;

                }, true);
            }

        });

        //Directive that contains d3 code to render proposed clonal evolution
        //Plot type: Fishbone
        tevApp.directive('figure4Plot', function(){
            return {
                restrict: 'A',
                scope: {
                    data: '=',
                    colors: '='
                },
                link: link
            };

            function link(scope, element){
                var data = scope.data;
                var el =  element[0];
                var width = el.clientWidth;
                var height = el.clientHeight;

                var color_ref = scope.colors;

                var cluster = d3.layout.cluster()
                        .size([420, 420]);

                //padding between independent fishplot backbones
                var fishbone_padding = 50;

                scope.$watchGroup(['data', 'colors'], function(updatedData){
                    d3.select(el).select('svg').remove();

                    data = updatedData[0];
                    color_ref = updatedData[1];
                    var additive_y = 50;

                    if(data.length == 0){
                        return;
                    }

                    function get_parent_clones(data){
                        var parent_indices = [];
                        for(var i = 0; i < data.length; i++){
                            if(data[i].parent_index_of_this == 'plot'){
                                parent_indices.push(i);
                            }
                        }
                        return parent_indices;
                    }

                    var independent_clonal_evos = get_parent_clones(data);

                    var svg = d3.select(el).append("svg")
                            .attr('id', 'fishbone_svg')
                            .attr("height", height)
                            .attr("width", width);

                    var title = svg.append('text')
                            .attr('class', 'title-text')
                            .attr('x', width/2)
                            .attr('y', 20)
                            .style('text-anchor', 'middle')
                            .text(function(){
                                return 'Fishbone of Suggested Fishplot';
                            });

                    for(var p = 0; p < independent_clonal_evos.length; p++) {

                        var fishbone_data = {};
                        fishbone_data.name = data[independent_clonal_evos[p]].key;
                        fishbone_data.children = [];
                        fishbone_data.cluster = data[independent_clonal_evos[p]].values[0].cluster;
                        fishbone_data.parent_allele = data[independent_clonal_evos[p]].values[0].parent_allele;

                        for (var i = 0; i < data[independent_clonal_evos[p]].children.length; i++) {
                            var child = data[independent_clonal_evos[p]].children[i];
                            child = data[child];
                            add_children(child, fishbone_data, data);
                        }

                        var cluster = d3.layout.cluster()
                                .size([100, 400]);

                        var diagonal = d3.svg.diagonal()
                                .projection(function (d) {
                                    return [d.y, d.x];
                                });

                        var fishbone = svg.append('svg')
                                .attr('width', width)
                                .attr('x', 0)
                                .attr('y', additive_y)
                                .append("g")
                                .attr('id', 'fishbone'+p)
                                .attr("transform", "translate(100," + 20 + ")");

                        var nodes = cluster.nodes(fishbone_data),
                                links = cluster.links(nodes);

                        var link = fishbone.selectAll(".link")
                                .data(links)
                                .enter().append("path")
                                .attr("class", "link")
                                .attr("d", diagonal);

                        var node = fishbone.selectAll(".node")
                                .data(nodes)
                                .enter().append("g")
                                .attr("class", "node")
                                .attr("transform", function (d) {
                                    return "translate(" + d.y + "," + d.x + ")";
                                });

                        node.append("circle")
                                .attr("r", 4.5)
                                .attr('fill', function (d) {
                                    return color_ref[d.parent_allele];
                                });

                        nodes.forEach(function (d) {
                            var names = d.name.split('\n');
                            for (var i = 0; i < names.length; i++) {
                                var name = names[i];
                                fishbone.append('text')
                                        .attr('class', 'node_text')
                                        .attr("transform", "translate(" + d.y + "," + d.x + ")")
                                        .attr('dx', function () {
                                            return d.children ? -8 : 8;
                                        })
                                        .attr('dy', function () {
                                            return ((i + 1) * 10);
                                        })
                                        .style("text-anchor", function () {
                                            return d.children ? "end" : "start";
                                        })
                                        .text(name);
                            }
                        });

                        additive_y = additive_y + d3.select('#fishbone'+p).node().getBBox().height + fishbone_padding;
                        if(additive_y > parseInt(d3.select(el).style('height'))){
                            d3.select(el).style('height', additive_y+'px');
                        }
                    }
                });

            }

        });

        //Directive that contains d3 code to render proposed clonal evolution and alternative allele freq
        //Plot type: Fishplot
        tevApp.directive('figure5Plot',['$window', function($window){

            return {
                restrict: 'A',
                scope: {
                    data: '=',
                    colors: '='
                },
                link: link
            };

            function link(scope, element){
                var data = scope.data;
                var el = element[0];
                var width = el.clientWidth;
                var height = el.clientHeight;
                var color_ref = scope.colors;

                scope.$watchGroup(['data', 'colors'], function(updatedData){
                    plotSuggestedFishplot(updatedData, el, height, width, color_ref, scope);

                    $window.addEventListener('resize', suggested_fishplot_resize);

                    function suggested_fishplot_resize(){
                        var new_width = d3.select(el).node().clientWidth;
                        plotSuggestedFishplot(updatedData, el, height, new_width, color_ref, scope);
                    }
                });
            }

        }]);

        tevApp.directive('fishplotEditor',['$http', '$filter', '$window', function ($http, $filter, $window) {
            return {
                restrict: 'A',
                scope: {
                    data: '=',
                    values: '=',
                    colors: '=',
                    chromosomes: '=',
                    positions: '=',
                    types: '=',
                    cdnas: '=',
                    refseqs: '=',
                    references: '=',
                    alternatives: '=',
                    patientswithgene: '='
                },
                link: link
            };

            function link(scope, element) {
                var data = scope.data;
                var el = element[0];
                var width = el.clientWidth;
                var height = el.clientHeight;
                var color_ref = scope.colors;
                var chromosome_ref = scope.chromosomes;
                var position_ref = scope.positions;
                var type_ref = scope.types;
                var cDNA_ref = scope.cdnas;
                var ref_seq_ref = scope.refseqs;
                var reference_ref = scope.references;
                var alternative_ref = scope.alternatives;
                var patients_with_gene_ref = scope.patientswithgene;

                //variables to hold canvas_nested_data and fishplot_svgs
                var resize_canvas_nested_data = [];
                var resize_fishplot_svgs = [];

                //When trying to get offsetTop for figure6 div it doesn't use heights of collapsed div
                //It uses their full heights, as if they were expanded
                var panel_heading_one_height = d3.select('#heading1').node().clientHeight;
                var panel_heading_two_height = d3.select('#heading2').node().clientHeight;
                var purity_div_height = d3.select('#purity-row').node().clientHeight;
                var estimated_accordian_padding = 3*20;
                var estimated_offsetTop = panel_heading_one_height+panel_heading_two_height+purity_div_height+estimated_accordian_padding;
                height = d3.select(window).node().outerHeight - estimated_offsetTop;

                //initial max VAF
                var max_VAF_value = 100;

                scope.$watch('colors', function(new_colors){
                    color_ref = new_colors;
                });

                scope.$watch('chromosomes', function(updated_info){
                    chromosome_ref = updated_info;
                });

                scope.$watch('positions', function(updated_info){
                    position_ref = updated_info;
                });

                scope.$watch('types', function(updated_info){
                    type_ref = updated_info;
                });

                scope.$watch('cdnas', function(updated_info){
                    cDNA_ref = updated_info;
                });

                scope.$watch('refseqs', function(updated_info){
                    ref_seq_ref = updated_info;
                });

                scope.$watch('references', function(updated_info){
                    reference_ref = updated_info;
                });

                scope.$watch('alternatives', function(updated_info){
                    alternative_ref = updated_info;
                });

                scope.$watch('patientswithgene', function(updated_info){
                    patients_with_gene_ref = updated_info;
                });

                scope.$watch('data', function(updated_data){
                    create_fishplot_editor(updated_data, el, height, width, color_ref, chromosome_ref, position_ref,
                    type_ref, cDNA_ref, ref_seq_ref, reference_ref, alternative_ref, patients_with_gene_ref, max_VAF_value,
                    scope, $window, $filter, $http);
                });
            }
        }]);

        $(function () {
            $('[data-toggle="popover"]').popover()
        });

        tevApp.directive('editorTable', function () {
            return {
                restrict: 'A',
                scope: {
                    data: '=',
                    colors: '=',
                    plotted: '=',
                    loadeddata: '='
                },
                link: link
            };

            function link(scope, element) {
                var el = element[0];
                var height = el.clientHeight;
                var width = el.clientWidth;
                var data = scope.data;
                var info = scope.values;

                var color_ref = scope.colors;
                var plotted = scope.plotted;

                var timepoints = [];

                scope.$watch('colors', function(new_colors){
                    color_ref = new_colors;
                });

                scope.$watch('plotted', function(updated_data){
                    plotted = updated_data;
                });

                scope.$watch('data', function(updated_data){
                    d3.selectAll('table').remove();
                    d3.selectAll('h2').remove();
                    data = updated_data;
                    var nested_data = d3.nest().key(function (d) {
                        return d.allele;
                    }).entries(data);

                    nested_data = sort_nested_alleles(nested_data);

                    timepoints = uniqueAxisLabels(data, 'Sample_Barcode');
                    timepoints = timepoints.sort(function(a, b){
                       return a-b;
                    });

                    var num_of_timepoints = timepoints.length;

                    var fishplot_data = [];

                    nested_data.forEach(function (d, i) {
                        var allele_data = {};
                        allele_data['allele'] = d.key;
                        allele_data['freqs'] = [];
                        var freqs = d.values;
                        for (var j = 0; j < freqs.length; j++) {
                            allele_data['freqs'].push(freqs[j].alt_count);
                        }
                        fishplot_data.push(allele_data);
                    });

                    add_conserved_alt_count(nested_data);


                    //Sorting data by what can fit in what
                    var sort_array = [];
                    for (var i = 0; i < fishplot_data.length; i++) {
                        //Have to rank when ordering, because a subclone that develops later
                        //cannot be parent of clone that derived earlier, even if it has a higher variant allele freq
                        var ranking_adjustment = Math.pow(10, num_of_timepoints);
                        var allele_freqs = fishplot_data[i].freqs;
                        for (var j = 0; j < allele_freqs.length; j++) {
                            if (allele_freqs[j] != 0) {
                                var rank = ranking_adjustment * allele_freqs[j];
                                sort_array.push({
                                    allele: fishplot_data[i].allele,
                                    rank: rank
                                });
                                break;
                            }
                            ranking_adjustment = ranking_adjustment / 10;
                        }
                    }

                    sort_array.sort(function (a, b) {
                        return b.rank - a.rank;
                    });

                    nested_data = order_by_rank(nested_data, sort_array);

                    //Any initial attributes on alleles will be applied within this for loop
                    for (var i = 0; i < nested_data.length; i++) {
                        nested_data[i].color = color_ref[nested_data[i].key];
                    }

                    d3.select(el).append('h2').style('text-align', 'center').html('VAF of Alleles at Each Timepoint');

                    var table = d3.select(el).append('table').attr('class', 'table table-bordered')
                            .attr('id', 'editor-table')
                            .attr('data-toggle', 'table')
                            .style('table-layout', 'fixed');

                    var header1 = table.append('thead').append('tr');
                    header1.append('th').html(' ');
                    header1.append('th').style('align', 'center').attr('colspan', timepoints.length).html('Timepoints');

                    var tbody = table.append('tbody').attr('id', 'tbody-editor-table').append('tr');
                    tbody.append('td').attr('align', 'center').html('<b> Alleles </b>');

                    for(var i = 0; i < timepoints.length; i++){
                        if(Number.isInteger(timepoints[i]) == true){
                            var timepoint = 'Day ' + timepoints[i];
                        }
                        else{
                            timepoint = timepoints[i];
                        }
                        tbody.append('td').attr('align', 'center').html('<b>' + timepoint + '</b>');
                    }

                    for(var i = 0; i < nested_data.length; i++){
                        var table_data = d3.select('#tbody-editor-table').append('tr').attr('id', 'tr'+i);
                        d3.select('#tr'+i).append('td')
                                .attr('bgcolor', nested_data[i].color)
                                .style('text-align', 'center')
                                .html(nested_data[i].key);
                        for (var j = 0; j < timepoints.length; j++) {
                            d3.select('#tr' + i).append('td')
                                    .style('text-align', 'center')
                                    .attr('value', nested_data[i].values[j].alt_count)
                                    .attr('conserved_val', nested_data[i].values[j].conserved_alt_count)
                                    .attr('timepoint', timepoints[j])
                                    .attr('allele', nested_data[i].key)
                                    .attr('index', i)
                                    .html(function(){
                                        if(nested_data[i].values[j].alt_count == nested_data[i].values[j].conserved_alt_count) {
                                            return (nested_data[i].values[j].alt_count + '%');
                                        }
                                        else{
                                            var conserved_alt_count = nested_data[i].values[j].conserved_alt_count;
                                            var html = '%' + ' '
                                                    + '<a data-toggle="popover" data-trigger="click" ' +
                                                    'title="This VAF has been altered"'
                                                    + 'data-content="The original VAF was ' + conserved_alt_count
                                                    + '">*</a>';
                                            return (nested_data[i].values[j].alt_count + html);
                                        }

                                    })
                                    .on('dblclick', function () {
                                        var index = parseInt(d3.select(this).attr('index'));
                                        if(plotted.indexOf(index) != -1){

                                            return;
                                        }

                                        var allele = d3.select(this).attr('allele');
                                        var alt_count = d3.select(this).attr('value');
                                        var conserved_alt_count = d3.select(this).attr('conserved_val');
                                        var current_td = d3.select(this);
                                        current_td.html(' ');
                                        var form = current_td.append('form');
                                        form.append('input')
                                                .attr('type', 'number')
                                                .attr('min', '0')
                                                .attr('max', '100')
                                                .attr('value', alt_count)
                                                .on('keydown', function () {
                                                    if (d3.event.keyCode == 13) {
                                                        d3.event.preventDefault();
                                                        if (!isNaN(parseInt(d3.select(this).node().value))) {
                                                            var new_vaf = parseInt(d3.select(this).node().value);
                                                            if (new_vaf > 100) {
                                                                new_vaf = 100;
                                                            }
                                                            if (new_vaf < 0) {
                                                                new_vaf = 0;
                                                            }
                                                            current_td.innerHTML = '';
                                                            var percentage = '%';
                                                            if(new_vaf != conserved_alt_count){
                                                                percentage = percentage + ' '
                                                                + '<a data-toggle="popover" data-trigger="click" ' +
                                                                        'title="This VAF has been altered"'
                                                                        + 'data-content="The original VAF was ' + conserved_alt_count
                                                                + '">*</a>';
                                                            }
                                                            current_td.attr('value', new_vaf).html(new_vaf + percentage);
                                                            var updated_info = [current_td.attr('allele'),
                                                                current_td.attr('timepoint'),
                                                                new_vaf];
                                                            //update with parent scope with info on the allele, timepoint,
                                                            //and new VAF
                                                            //updated_editor_info will be used to communicate between the table
                                                            //and the fishplot editor
                                                            scope.$apply(function () {
                                                                scope.$parent.updated_editor_info = updated_info;
                                                            });
                                                        }
                                                    }
                                                });
                                        form.append('label').html('%');
                                    });
                        }
                    }

                });

                scope.$watch('loadeddata', function(loaded_data){
                    var nested_data = JSON.parse(JSON.stringify(loaded_data));

                    //check if origin already added. If so, remove it
                    for(var i = 0; i < nested_data.length; i++){
                        var timepoint_data = nested_data[i].values;
                        if(timepoint_data[0].Sample_Barcode < 0 || timepoint_data[0].Sample_Barcode == 'origin'){
                            timepoint_data.shift();
                        }
                    }

                    d3.select('#tbody-editor-table').remove();
                    var tbody = d3.select('#editor-table').append('tbody').attr('id', 'tbody-editor-table').append('tr');
                    tbody.append('td').attr('align', 'center').html('<b> Alleles </b>');

                    for(var i = 0; i < timepoints.length; i++){
                        if(Number.isInteger(timepoints[i]) == true){
                            var timepoint = 'Day ' + timepoints[i];
                        }
                        else{
                            timepoint = timepoints[i];
                        }
                        tbody.append('td').attr('align', 'center').html('<b>' + timepoint + '</b>');
                    }

                    for(var i = 0; i < nested_data.length; i++){
                        var table_data = d3.select('#tbody-editor-table').append('tr').attr('id', 'tr'+i);
                        d3.select('#tr'+i).append('td')
                                .attr('bgcolor', nested_data[i].color)
                                .style('text-align', 'center')
                                .html(nested_data[i].key);
                        for (var j = 0; j < timepoints.length; j++) {
                            d3.select('#tr' + i).append('td')
                                    .style('text-align', 'center')
                                    .attr('value', nested_data[i].values[j].alt_count)
                                    .attr('conserved_val', nested_data[i].values[j].conserved_alt_count)
                                    .attr('timepoint', timepoints[j])
                                    .attr('allele', nested_data[i].key)
                                    .attr('index', i)
                                    .html(function(){
                                        if(nested_data[i].values[j].alt_count == nested_data[i].values[j].conserved_alt_count) {
                                            return (nested_data[i].values[j].alt_count + '%');
                                        }
                                        else{
                                            var conserved_alt_count = nested_data[i].values[j].conserved_alt_count;
                                            var html = '%' + ' '
                                                    + '<a data-toggle="popover" data-trigger="click" ' +
                                                    'title="This VAF has been altered"'
                                                    + 'data-content="The original VAF was ' + conserved_alt_count
                                                    + '">*</a>';
                                            return (nested_data[i].values[j].alt_count + html);
                                        }

                                    })
                                    .on('dblclick', function () {
                                        var index = parseInt(d3.select(this).attr('index'));
                                        if(plotted.indexOf(index) != -1){

                                            return;
                                        }

                                        var allele = d3.select(this).attr('allele');
                                        var alt_count = d3.select(this).attr('value');
                                        var conserved_alt_count = d3.select(this).attr('conserved_val');
                                        var current_td = d3.select(this);
                                        current_td.html(' ');
                                        var form = current_td.append('form');
                                        form.append('input')
                                                .attr('type', 'number')
                                                .attr('min', '0')
                                                .attr('max', '100')
                                                .attr('value', alt_count)
                                                .on('keydown', function () {
                                                    if (d3.event.keyCode == 13) {
                                                        d3.event.preventDefault();
                                                        if (!isNaN(parseInt(d3.select(this).node().value))) {
                                                            var new_vaf = parseInt(d3.select(this).node().value);
                                                            if (new_vaf > 100) {
                                                                new_vaf = 100;
                                                            }
                                                            if (new_vaf < 0) {
                                                                new_vaf = 0;
                                                            }
                                                            current_td.innerHTML = '';
                                                            var percentage = '%';
                                                            if(new_vaf != conserved_alt_count){
                                                                percentage = percentage + ' '
                                                                + '<a data-toggle="popover" data-trigger="click" ' +
                                                                        'title="This VAF has been altered"'
                                                                        + 'data-content="The original VAF was ' + conserved_alt_count
                                                                + '">*</a>';
                                                            }
                                                            current_td.attr('value', new_vaf).html(new_vaf + percentage);
                                                            var updated_info = [current_td.attr('allele'),
                                                                current_td.attr('timepoint'),
                                                                new_vaf];
                                                            //update with parent scope with info on the allele, timepoint,
                                                            //and new VAF
                                                            //updated_editor_info will be used to communicate between the table
                                                            //and the fishplot editor
                                                            scope.$apply(function () {
                                                                scope.$parent.updated_editor_info = updated_info;
                                                            });
                                                        }
                                                    }
                                                });
                                        form.append('label').html('%');
                                    });
                        }
                    }
                });
            }

        });

        tevApp.directive('purityLinePlot',['$window', function($window){

            return {
                restrict: 'A',
                scope: {
                    data: '='
                },
                link: link
            };

            function link(scope, element){
                var data = scope.data;
                var el = element[0];
                var width = el.clientWidth;
                var height = el.clientHeight;

                scope.$watch('data', function(updated_data){
                    makePurityLinePlot(updated_data, el, height, width);
                    $window.addEventListener('resize', purity_resize);
                    function purity_resize() {
                        var new_height = d3.select(el).node().clientHeight;
                        var new_width = d3.select(el).node().clientWidth;
                        makePurityLinePlot(updated_data, el, new_height, new_width);
                    };
                });

            }

        }]);

        $('.closeall').ready(function(){
            $('.panel-collapse.in')
            .collapse('hide');
        });


    })();

    function change_icon(icon_number){
        var icon_id = '#expand'+icon_number;
        var is_expanded = d3.select(icon_id).attr('expanded');
        if(is_expanded == 'false'){
            d3.select(icon_id).attr('class', 'fa fa-minus');
            d3.select(icon_id).attr('expanded', 'true');
            return;
        }
        if(is_expanded == 'true'){
            d3.select(icon_id).attr('class', 'fa fa-plus');
            d3.select(icon_id).attr('expanded', 'false');
            return;
        }

    }

    function update_clonal_information_text(){
        var fishplot_index = d3.select('#ploidy_input').attr('fishplot_index');

        d3.select('#ploidy_text'+fishplot_index)
                .text(function(){
                   return 'Ploidy: ' + d3.select('#ploidy_input').node().valueAsNumber;
                });
    }

    function fetch_ploidy() {
        d3.select('#context-div').style('opacity', 0).attr('cm-showing', 'false');
        var index = parseInt(d3.select('#edit-ploidy-btn').attr('fish-index'));
        var ploidy_num = d3.select('#ploidy_text' + index).node().innerHTML.split(' ')[1];
        d3.select('#ploidy_input').node().value = ploidy_num;
        d3.select('#ploidy_input').attr('fishplot_index', index);

        $("#exampleModal").modal('show');
    }

</script>
<body ng-controller = "tevDataController" style="margin-left: 2.5%; margin-right: 2.5%; margin-bottom: 2.5%">
<!--
Dont need to provide alleles to figure 1 because all that's changing is the colors
-->

<nav class="navbar navbar-inverse navbar-fixed-top" role="navigation">
    <div class="container">
        <!-- Brand and toggle get grouped for better mobile display -->
        <div class="navbar-header">
            <button type="button" class="navbar-toggle" data-toggle="collapse"
                    data-target="#bs-example-navbar-collapse-1">
                <span class="sr-only">Toggle navigation</span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
                <span class="icon-bar"></span>
            </button>
            <a id="subject_brand" class="navbar-brand" href=""></a>
        </div>
        <!-- Collect the nav links, forms, and other content for toggling -->
        <div class="collapse navbar-collapse" id="bs-example-navbar-collapse-1">
            <ul class="nav navbar-nav">
                <li class="active">
                    <a data-toggle="tab" href="#editor-div">Fishplot Editor</a>
                </li>
                <li>
                    <a data-toggle="tab" href="#summary-divs">Variant Allele Summary</a>
                </li>
            </ul>
            <ul class="nav navbar-nav navbar-right">
                <li>
                    <a href="../../data/">Return to Data</a>
                </li>
            </ul>
        </div>
        <!-- /.navbar-collapse -->
    </div>
    <!-- /.container -->
</nav>
<div class="tab-content">
    <div id="editor-div" class="tab-pane fade in active">
        <div class="row">
            <div id="editor-wrapper" class="col-lg-12 col-md-12 col-sm-12">
                <div id="collapsable-row" class="row">
                    <div class="panel-group allclose" id="accordion" role="tablist" aria-multiselectable="true">
                        <div class="panel panel-default">
                            <div id="heading1" class="panel-heading panel-heading-custom" role="tab" id="headingOne" style="background: #222222;">
                                <h4 class="panel-title">
                                    <a id="expand-btn-1" role="button" onclick="change_icon(1)" data-toggle="collapse"
                                       href="#collapseOne"
                                       aria-expanded="true" aria-controls="collapseOne"
                                       style="color: #ffffff;">
                                        <i id="expand1" expanded=false class="fa fa-plus" aria-hidden="true"></i> VAF Over Time
                                    </a>
                                </h4>
                            </div>
                            <div id="collapseOne" class="panel-collapse collapse in" role="tabpanel"
                                 aria-labelledby="headingOne">
                                <div class="panel-body">
                                    <div id="figure1" figure1-plot data="clustered_alleleified_data" colors="color_ref"
                                         style="height: 600px; width: 100%; float:left;">
                                    </div>
                                </div>
                            </div>
                        </div>
                        <div class="panel panel-default">
                            <div id="heading2" class="panel-heading panel-heading-custom" role="tab" id="headingTwo" style="background: #222222;">
                                <h4 class="panel-title">
                                    <a id="expand-btn-2" onclick="change_icon(2)" class="collapsed" role="button" data-toggle="collapse"
                                       href="#collapseTwo" aria-expanded="false" aria-controls="collapseTwo"
                                        style="color: #ffffff;">
                                        <i id="expand2" expanded=false class="fa fa-plus" aria-hidden="true"></i> VAF Editor Table
                                    </a>
                                </h4>
                            </div>
                            <div id="collapseTwo" class="panel-collapse collapse" role="tabpanel"
                                 aria-labelledby="headingTwo">
                                <div class="panel-body">
                                    <div id="figure7" editor-table data="alleleified_data" colors="color_ref"
                                         plotted="plotted"
                                         loadeddata="loaded_fishplot_data"
                                         style="width: 100%; float: left;">
                                    </div>
                                </div>
                            </div>
                        </div>
                    </div>
                </div>

                <div id="purity-row" class="row">
                    <div id="purity-line-div" purity-line-plot data="purity_data" class="col-sm-12 col-md-12 col-lg-12"
                         style="height: 150px; float: left;">
                    </div>
                </div>

                <div id="editor-row" class="row" >
                <div id="figure6-wrapper" style="overflow: visible;">
                <div id="figure6" fishplot-editor data="alleleified_data" values="updated_editor_info" colors="color_ref"
                     chromosomes="chromosome_dict" positions="position_dict" types="type_dict" cdnas="cDNA_dict" refseqs="ref_seq_dict"
                     references="reference_dict" alternatives="alternative_dict" patientswithgene="patient_gene_dict"
                     style="width: 100%; height: 600px; float: left; overflow: hidden;">
                </div>
                </div>
            </div>

            <div id="context-div" cm-showing="false" class="btn-group-vertical" style="z-index: -1; opacity: 0; position: absolute;">
                <button id="rmv-fishplot-btn" type="button" class="btn btn-default">Remove fishplot(s)</button>
                <button id="edit-ploidy-btn" type="button" class="btn btn-default" onclick="fetch_ploidy()">Edit ploidy</button>
            </div>

            <!--
            Modal for editing ploidy
            -->
            <div class="modal fade bd-example-modal-sm" id="exampleModal" tabindex="-1" role="dialog" aria-labelledby="exampleModalLabel"
                 aria-hidden="true">
                <div class="modal-dialog modal-sm" role="document">
                    <div class="modal-content">
                        <div class="modal-header modal-header-inverse">
                            <button type="button" class="close" data-dismiss="modal" aria-label="Close" style="color: #ffffff;">
                                <span aria-hidden="true">&times;</span>
                            </button>
                            <h4 class="modal-title" id="exampleModalLabel">Edit Clonal Information</h4>
                        </div>
                        <div class="modal-body">
                            <form>

                                <label for="ploidy_input" class="form-control-label">Ploidy:</label>
                                <input class="form-control" id="ploidy_input" type="number" min="0" max="100"/>


                            </form>
                        </div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-default" data-dismiss="modal" id="save_button"
                                    onclick="update_clonal_information_text()">Save
                            </button>
                        </div>
                    </div>
                </div>
            </div>

            <!--
            Modal for editing purity at timepoints
            -->

            <div class="modal fade bs-example-modal-sm" id="purityModal" tabindex="-1" role="dialog" aria-labelledby="mySmallModalLabel"
                 aria-hidden="true">
                <div class="modal-dialog modal-sm" role="document">
                    <div class="modal-content">
                        <div class="modal-header modal-header-inverse">
                            <button type="button" class="close" data-dismiss="modal" aria-label="Close" style="color: #ffffff;">
                                <span aria-hidden="true">&times;</span>
                            </button>
                            <h4 class="modal-title" id="purityModalHeader">Edit Purity Information</h4>
                        </div>
                        <div class="modal-body">
                            <form id="purityForm" >

                                <!--
                                Will edit the form in d3 based on the timepoints
                                -->

                            </form>
                        </div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-default" data-dismiss="modal"
                                    id="save_purity_button">
                                Save
                            </button>
                        </div>
                    </div>
                </div>
            </div>

            <!--
            Modal for selecting the CCF
            -->

            <div class="modal fade" id="CCF_modal" tabindex="10" role="dialog" aria-labelledby="exampleModalLabel"
                 aria-hidden="true" style="overflow: auto;">
                <div class="modal-dialog" role="document">
                    <div class="modal-content">
                        <div class="modal-header modal-header-inverse">
                            <button type="button" class="close" data-dismiss="modal" aria-label="Close" style="color: #ffffff;">
                                <span aria-hidden="true">&times;</span>
                            </button>
                            <h4 class="modal-title" id="exampleModalLabel">Choose CCF Distribution</h4>
                        </div>
                        <div id="CCF_modal_body" class="modal-body" popped-up="0">
                            <!--
                            Using D3 in the fishplot editor div, a table will be filled in here
                            -->
                        </div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-default" data-dismiss="modal" id="save-CCF-button">Choose
                            </button>
                        </div>
                    </div>
                </div>
            </div>

            <!--
            Modal for selecting the CCF during reprompt
            -->

            <div class="modal fade" id="CCF_modal_reprompt" tabindex="10" role="dialog" aria-labelledby="exampleModalLabel"
                 aria-hidden="true" style="overflow: auto;">
                <div class="modal-dialog" role="document">
                    <div class="modal-content">
                        <div class="modal-header modal-header-inverse">
                            <button type="button" class="close" data-dismiss="modal" aria-label="Close" style="color: #ffffff;">
                                <span aria-hidden="true">&times;</span>
                            </button>
                            <h4 class="modal-title" id="exampleModalLabel">Choose CCF Distribution</h4>
                        </div>
                        <div id="CCF_modal_body_reprompt" class="modal-body">
                            <!--
                            Using D3 in the fishplot editor div, a table will be filled in here
                            -->
                        </div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-default" data-dismiss="modal" id="save-CCF-button-reprompt">Choose
                            </button>
                        </div>
                    </div>
                </div>
            </div>

            <!--
            Modal to prompt "save as:" for fishplot editor
            -->

            <div class="modal fade" id="save_fishplot_modal" tabindex="-1" role="dialog"
                 aria-labelledby="exampleModalLabel"
                 aria-hidden="true">
                <div class="modal-dialog" role="document">
                    <div class="modal-content">
                        <div class="modal-header modal-header-inverse">
                            <button type="button" class="close" data-dismiss="modal" aria-label="Close" style="color: #ffffff;">
                                <span aria-hidden="true">&times;</span>
                            </button>
                            <h4 class="modal-title" id="exampleModalLabel">Save Your Fishplot</h4>
                        </div>
                        <div class="modal-body">
                            <form>

                                <label for="save_fishplot_as" class="form-control-label">Save as: </label>
                                <input class="form-control" id="save_fishplot_as" type="text"/>

                            </form>
                        </div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-default" data-dismiss="modal"
                                    id="save_as_fishplot_button">Save
                            </button>
                        </div>
                    </div>
                </div>
            </div>

            <!--
            Modal to select and load saved fishplots
            -->

            <div class="modal fade" id="get_fishplot_modal" tabindex="-1" role="dialog"
                 aria-labelledby="exampleModalLabel" aria-hidden="true">
                <div class="modal-dialog" role="document">
                    <div class="modal-content">
                        <div class="modal-header modal-header-inverse">
                            <button type="button" class="close" data-dismiss="modal" aria-label="Close" style="color: #ffffff;">
                                <span aria-hidden="true">&times;</span>
                            </button>
                            <h4 class="modal-title" id="exampleModalLabel">Select Saved Fishplot</h4>
                        </div>
                        <div id="get-fishplot-modal-body" class="modal-body">

                            <!--
                            Where the saved fishplot selection dropdown will be added via D3
                            -->

                        </div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-default" data-dismiss="modal"
                                    id="get_fishplot_button">Choose
                            </button>
                        </div>
                    </div>
                </div>
            </div>

            <!--
            Modal to notify the user wheither or not the fishplot was saved successfully
            -->

            <div class="modal fade" id="saved-fishplot-dialog-box" role="dialog">
                <div class="modal-dialog modal-md">
                    <div class="modal-content">
                        <div class="modal-header modal-header-inverse" >
                            <button type="button" class="close" data-dismiss="modal" style="color: #ffffff;">&times;</button>
                            <h4 id="dialog-header" class="modal-title" style="color: #ffffff;"></h4>
                        </div>
                        <div id="dialog-body" class="modal-body">

                            <!--
                            Will fill this in with a custom message pertaining to the name the user saved the fishplot under
                            -->

                        </div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-default btn-sm" data-dismiss="modal">OK</button>
                        </div>
                    </div>
                </div>
            </div>

             <!--
            Modal for information on variant/gene
            -->

            <div class="modal fade" id="gene-info-modal" role="dialog">
                <div class="modal-dialog modal-md">
                    <div class="modal-content">
                        <div class="modal-header modal-header-inverse">
                            <button type="button" class="close" data-dismiss="modal" style="color: #ffffff;">&times;</button>
                            <h4 id="gene-info-modal-header-title" class="modal-title" style="color: #ffffff; text-align:center;"></h4>
                        </div>
                        <div id="gene-info-modal-body" class="modal-body">

                            <!--
                            Will fill this in with information regarding gene
                            -->

                        </div>
                        <div class="modal-footer">
                            <button type="button" class="btn btn-default btn-sm" data-dismiss="modal">OK</button>
                        </div>
                    </div>
                </div>
            </div>


        </div>
    </div>
</div>


    <div id="summary-divs" class="tab-pane">
        <div class="panel panel-primary">
            <div id="summary-stats-panel-heading" class="panel-heading" style="background: #222222; display: block; border-color: #222222;">
                <h4 style="text-align: center; margin: 0px;">Summary Statistics</h4>
            </div>
            <div class="panel-body" style="padding: 0px;">
                <div class="row col-lg-12">
                    <div class="col-lg-6 col-md-6 col-sm-12">
                        <div id="figure1" figure1-plot data="clustered_alleleified_data" colors="color_ref"
                             class="col-lg-11 col-md-11 col-sm-11" style="height: 650px; float:left;">
                        </div>
                    </div>

                    <div class="col-lg-6 col-md-6 col-sm=12" style="float: right;">
                        <div id="figure2" figure2-plot data="[clustered_alleleified_data, alleles]" colors="color_ref"
                             class="col-lg-11 col-md-11 col-sm-11" style="height: 650px; float: right;">
                        </div>
                    </div>
                </div>


                <div class="row col-lg-12">
                    <div class="col-lg-6">
                        <div id="figure3" figure3-plot data="[dendro_data, dendro_y_data, alleleified_data, color_ref]"
                             class="col-lg-11"
                             style="height: 600px; float: left;">
                        </div>
                    </div>

                    <div class="col-lg-6">
                        <div id="figure4" figure4-plot data="fishbone_data" colors="color_ref" class="col-lg-12"
                             style="height: 600px; float: right; padding: 0px;">
                        </div>
                    </div>
                </div>


                <div class="row">
                    <div class="col-lg-12 col-md-12 col-sm-12">
                        <div id="figure5" figure5-plot data="linked_alleleified_data" colors="color_ref"
                             style="width: 100%; height: 600px; float: left;">
                        </div>
                    </div>
                </div>
            </div>
        </div>
    </div>
</div>

</body>
</html>